alibi/explainers/counterfactual_tf2.py

```diff
@@ -0,0 +1,1028 @@
+            features, but can be passed an array for feature-wise step sizes.
+        decay
+            Flag to decay learning rate to zero for each outer loop over lambda.
+
+        Returns
+        -------
+        *explanation* - a dictionary containing the counterfactual with additional metadata.
+
+        """  # noqa W605
+
+        # TODO: ALEX: DOCUMENT SHAPE REQUIREMENTS for `X` in docstring above
+        # TODO: ALEX: Improve docstring for `early_stopping`  and `tol` above?
+        # TODO: EXPLAIN WHEN `eps` is USED above?
+
+        if X.shape[0] != 1:
+            logger.warning('Currently only single instance explanations supported (first dim = 1), '
+                           'but first dim = %s', X.shape[0])
+        # TODO: ALEX: CHECK TENSOR SHAPES HERE AND UPDATE ACCORDINGLY
+        self.instance_numpy = X
+        self.target_proba_numpy = np.array([target_proba])[:, np.newaxis]
+
+        # update options for the optimiser
+        self.decay = decay
+        self.learning_rate_init = learning_rate_init
+        self.lr_schedule = learning_rate_init
+        self.max_iter = max_iter
+        self._initialise_optimiser()
+
+        # make a prediction
+        Y = self.make_prediction(X)
+        instance_class = self._get_class(to_numpy_arr(Y))
+        instance_proba = to_numpy_arr(Y[:, instance_class])
+
+        # helper function to return the model output given the target class
+        self.get_cf_prediction = partial(self._get_cf_prediction, instance_class=instance_class)
+        # initialize optimised variables, targets and loss weight
+        self._initialise_variables(X, target_class, target_proba)
+        lam_dict = self._initialise_lam()
+
+        # search for a counterfactual by optimising loss starting from the original solution
+        result = self._search(init_cf=X, init_lam=lam_dict['midpoint'], init_lb=lam_dict['lb'], init_ub=lam_dict['ub'])
+        self._reset_step()
+
+        return self.build_explanation(X, result, instance_class, instance_proba)
+
+    def make_prediction(self, X: Union[np.ndarray, tf.Variable, tf.Tensor]) -> tf.Tensor:
+        """
+        Makes a prediction for data points in `X`
+        """
+        # TODO: ALEX: DOCSTRING
+
+        return self.predictor(X, training=False)
+
+    def _get_class(self, Y: np.ndarray, threshold: float = 0.5) -> int:
+        """
+        Given a prediction vector containing the model output for a prediction on a single data point,
+        it returns an integer representing the predicted class of the data point. Accounts for the case
+        where a binary classifier returns a single output.
+        """
+
+        # TODO: ALEX: TBD: Parametrise `threhold` in explain or do we assume people always assign labels on this rule?
+        # TODO: ALEX: ADD DOCSTRING
+
+        if Y.shape[1] == 1:
+            return int(Y > threshold)
+        else:
+            return np.argmax(Y)
+
```

Comment:
Might be missing edge case where `Y` has shape `(n,)` for custom models.

Reply:
So if you give a tensor of dimension [B, N] to a predictor in TF, do you not get [B, X] where X is output dimension? The docs clearly state there should be a leading singleton dimension and an error is raised if that's not the case.